src/native/managed/cdacreader/src/Contracts/Metadata_1.cs

```diff
@@ -0,0 +1,333 @@
+                MTFlags = _target.Read<uint>(Address + (ulong)_type.Fields[nameof(Metadata_1.MethodTableFlags.MTFlags)].Offset),
+                MTFlags2 = _target.Read<uint>(Address + (ulong)_type.Fields[nameof(Metadata_1.MethodTableFlags.MTFlags2)].Offset),
+                BaseSize = _target.Read<uint>(Address + (ulong)_type.Fields[nameof(Metadata_1.MethodTableFlags.BaseSize)].Offset),
+            };
+            _methodTableFlags = flags;
+        }
+        return _methodTableFlags.Value;
+    }
+
+    internal Metadata_1.MethodTableFlags Flags => GetOrCreateFlags();
+
+    internal TargetPointer EEClassOrCanonMT => _target.ReadPointer(Address + (ulong)_type.Fields[nameof(EEClassOrCanonMT)].Offset);
+    internal TargetPointer EEClass => Metadata_1.GetEEClassOrCanonMTBits(EEClassOrCanonMT) == Metadata_1.EEClassOrCanonMTBits.EEClass ? EEClassOrCanonMT : throw new InvalidOperationException("not an EEClass");
+    internal TargetPointer CanonMT
+    {
+        get
+        {
+            if (Metadata_1.GetEEClassOrCanonMTBits(EEClassOrCanonMT) == Metadata_1.EEClassOrCanonMTBits.CanonMT)
+            {
+                return new TargetPointer((ulong)EEClassOrCanonMT & ~(ulong)Metadata_1.EEClassOrCanonMTBits.Mask);
+            }
+            else
+            {
+                throw new InvalidOperationException("not a canonical method table");
+            }
+        }
+    }
+}
+
+internal struct NonValidatedEEClass_1
+{
+    public readonly Target _target;
+    private readonly Target.TypeInfo _type;
+
+    internal TargetPointer Address { get; init; }
+
+    internal NonValidatedEEClass_1(Target target, TargetPointer eeClassPointer)
+    {
+        _target = target;
+        Address = eeClassPointer;
+        _type = target.GetTypeInfo(DataType.EEClass);
+    }
+
+    internal TargetPointer MethodTable => _target.ReadPointer(Address + (ulong)_type.Fields[nameof(MethodTable)].Offset);
+}
+
+
+internal struct MethodTable_1
+{
+    internal Metadata_1.MethodTableFlags Flags { get; }
+    internal ushort NumInterfaces { get; }
+    internal ushort NumVirtuals { get; }
+    internal TargetPointer ParentMethodTable { get; }
+    internal TargetPointer Module { get; }
+    internal TargetPointer EEClassOrCanonMT { get; }
+    internal MethodTable_1(Data.MethodTable data)
+    {
+        Flags = new Metadata_1.MethodTableFlags
+        {
+            MTFlags = data.MTFlags,
+            MTFlags2 = data.MTFlags2,
+            BaseSize = data.BaseSize,
+        };
+        NumInterfaces = data.NumInterfaces;
+        NumVirtuals = data.NumVirtuals;
+        EEClassOrCanonMT = data.EEClassOrCanonMT;
+        Module = data.Module;
+        ParentMethodTable = data.ParentMethodTable;
+    }
+}
+
+internal partial struct Metadata_1 : IMetadata
+{
+    private readonly Target _target;
+    private readonly TargetPointer _freeObjectMethodTablePointer;
+
+    // FIXME: we mutate this dictionary - copies of the Metadata_1 struct share this instance
```

Comment:
Can we use `ReadOnlyDictionary` instead?

Reply:
We _want_ to mutate the dictionary: after a putative method table address has been validated, we work with instnaces from the dictionary.